src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -184,10 +130,549 @@ def _validate_certificate_with_key(certificate, private_key, schema_name, verror
             'Please provide a valid csr_id which has a valid CSR filed'
         )
 
-    await middleware.run_in_thread(
-        _validate_certificate_with_key, certificate, private_key, schema_name, verrors
+    await middleware.call(
+        'cryptokey.validate_certificate_with_key', certificate, private_key, schema_name, verrors, passphrase
     )
 
+    key_type = data.get('key_type')
+    if key_type:
+        if key_type != 'EC' and not data.get('key_length'):
+            verrors.add(
+                f'{schema_name}.key_length',
+                'This field is required with RSA based keys'
+            )
+
+
+class CryptoKeyService(Service):
+
+    ec_curve_default = 'BrainpoolP384R1'
+
+    ec_curves = [
+        'BrainpoolP512R1',
+        'BrainpoolP384R1',
+        'BrainpoolP256R1',
+        'SECP256K1'
```

Comment:
I dont know much about EC but do you care explain why you choose to implement these specific curves?

Reply:
Firstly it was aimed at not to overwhelm the users with the possibility of so many curves. Secondly, brain pool curves are slightly slower then the other NIST ones but they offer better security against some attacks. So I added another NIST secp* curve which should be fast comparatively. I was thinking we can gauge user acceptance and see if they would like us to add more curves, we can do so easily then. Otherwise we can add all the curves which cryptography supports. Thoughts ?